lib/srv/discovery/access_graph_azure.go

```diff
@@ -0,0 +1,409 @@
+					},
+				},
+			},
+		)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+	}
+	return nil
+}
+
+// azurePush upserts and deletes Azure resources to/from the Access Graph
+func azurePush(
+	client accessgraphv1alpha.AccessGraphService_AzureEventsStreamClient,
+	upsert *accessgraphv1alpha.AzureResourceList,
+	toDel *accessgraphv1alpha.AzureResourceList,
+) error {
+	err := azurePushUpsertInBatches(client, upsert)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	err = azurePushDeleteInBatches(client, toDel)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	err = client.Send(
+		&accessgraphv1alpha.AzureEventsStreamRequest{
+			Operation: &accessgraphv1alpha.AzureEventsStreamRequest_Sync{},
+		},
+	)
+	return trace.Wrap(err)
+}
+
+// getAllTAGSyncAzureFetchers returns both static and dynamic TAG Azure fetchers
+func (s *Server) getAllTAGSyncAzureFetchers() []*azuresync.Fetcher {
+	allFetchers := make([]*azuresync.Fetcher, 0, len(s.dynamicTAGAzureFetchers))
+
+	s.muDynamicTAGAzureFetchers.RLock()
+	for _, fetcherSet := range s.dynamicTAGAzureFetchers {
+		allFetchers = append(allFetchers, fetcherSet...)
+	}
+	s.muDynamicTAGAzureFetchers.RUnlock()
+
+	allFetchers = append(allFetchers, s.staticTAGAzureFetchers...)
+	return allFetchers
+}
+
+// initializeAndWatchAzureAccessGraph initializes and watches the TAG Azure stream
+func (s *Server) initializeAndWatchAzureAccessGraph(ctx context.Context, reloadCh chan struct{}) error {
+	// Check if the access graph is enabled
+	clusterFeatures := s.Config.ClusterFeatures()
+	policy := modules.GetProtoEntitlement(&clusterFeatures, entitlements.Policy)
+	if !clusterFeatures.AccessGraph && !policy.Enabled {
+		return trace.Wrap(errTAGFeatureNotEnabled)
+	}
+
+	// Configure the access graph semaphore for constraining multiple discovery servers
+	const (
+		semaphoreExpiration = time.Minute
+		semaphoreName       = "access_graph_azure_sync"
+		serviceConfig       = `{
+		 "loadBalancingPolicy": "round_robin",
+		 "healthCheckConfig": {
+			 "serviceName": ""
+		 }
+	 }`
+	)
+	lease, err := services.AcquireSemaphoreLockWithRetry(
+		ctx,
+		services.SemaphoreLockConfigWithRetry{
+			SemaphoreLockConfig: services.SemaphoreLockConfig{
+				Service: s.AccessPoint,
+				Params: types.AcquireSemaphoreRequest{
+					SemaphoreKind: types.KindAccessGraph,
+					SemaphoreName: semaphoreName,
+					MaxLeases:     1,
+					Holder:        s.Config.ServerID,
+				},
+				Expiry: semaphoreExpiration,
+				Clock:  s.clock,
+			},
+			Retry: retryutils.LinearConfig{
+				Clock:  s.clock,
+				First:  time.Second,
+				Step:   semaphoreExpiration / 2,
+				Max:    semaphoreExpiration,
+				Jitter: retryutils.DefaultJitter,
+			},
+		},
+	)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	ctx, cancel := context.WithCancel(lease)
+	defer cancel()
+	defer func() {
+		lease.Stop()
+		if err := lease.Wait(); err != nil {
+			s.Log.WarnContext(ctx, "error cleaning up semaphore", "error", err)
+		}
```

Comment:
Logging the semaphore kind and name can be useful as well (I'm not sure if it's already included in the `err`)